### PR TITLE
Feat/peer discovery

### PR DESCRIPTION
**Pull request by rahulghangas**
This PR adds ability to discover new peers in airwave by sending requests for information to known peers

- [x] Send Ping Messages
- [x] `DidReceivePing`
- [x] `DidReceivePingAck`
- [x] Correctness Tests

### CODE REVIEW

File: peer/peer.go [hunk 1]
```diff
@@ -80,8 +82,16 @@ func (p *Peer) Gossip(ctx context.Context, contentID []byte, subnet *id.Hash) {
 	p.gossiper.Gossip(ctx, contentID, subnet)
 }
 
+func (p *Peer) PeerDiscovery(ctx context.Context) {
+	p.discoveryClient.DiscoverPeers(ctx)
+}
```
Comment (loongy) [inline peer/peer.go:87]: This method should not exist. Instead, we should run `DiscoverPeers` during the `Peer.Run` method, because `DiscoverPeers` is a method that runs constantly in the background. Then, the `Peer.Run` function would look like something like this:

```
	// ... other stuff

	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		p.discoveryClient.DiscoverPeers(ctx)
	}()
	go func() {
		defer wg.Done()
		p.transport.Run(ctx)
	}()
	wg.Wait()
```

File: peer/peerdiscovery.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+func (dc *DiscoveryClient) DiscoverPeers(ctx context.Context) {
+	var maxExpectedPeers int
+	if dc.opts.MaxExpectedPeers <= 0  {
+		dc.opts.Logger.Debug("pinging", zap.String("max expected peers cannot be less than 1", "setting to default - 5"))
+		maxExpectedPeers = 5
+	} else {
+		maxExpectedPeers = dc.opts.MaxExpectedPeers
+	}
+
+	go func() {
```
Comment (loongy) [inline peer/peerdiscovery.go:67]: This go routine should not exist. It is better to give control to the caller.

File: peer/peerdiscovery.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		for {
+			for _, sig := range dc.transport.Table().Peers(dc.opts.Alpha) {
+				msg.To = id.Hash(sig)
+				err := dc.transport.Send(ctx, sig, msg)
+				if err != nil {
+					dc.opts.Logger.Debug("pinging", zap.Error(err))
+					dc.transport.Table().DeletePeer(sig)
+				}
+			}
+			<-time.After(dc.opts.PingTimePeriod)
```
Comment (loongy) [inline peer/peerdiscovery.go:86]: You should select on the context here, because it is reasonable to expect that PingTimePeriod is very long.

```
		select {
		case <-ctx.Done():
			return
		case <-time.After(dc.opts.PingTimePeriod):
		}
```

Comment (loongy) [reply peer/peerdiscovery.go:86]: Also, consider the different between absolute and relative delays. You have a relative delay here (no matter how long the actual pinging takes, you will wait an additional PingTimePeriod). If you want an absolute time delay (always ping exactly once per PingTimePeriod), then you should use a ticker.

File: peer/peerdiscovery.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		To:      id.Hash(from),
+		Data:    addressByteSlice,
+	}
+	err := dc.transport.Send(ctx, from, response)
+	if err != nil {
+		dc.opts.Logger.Debug("sending pingAck", zap.Error(err))
+	}
+}
+
+func (dc *DiscoveryClient) didReceivePingAck(from id.Signatory, msg wire.Msg) {
```
Comment (loongy) [inline peer/peerdiscovery.go:142]: Surge is able to unmarshal directly into a slice. For example:

```
slice := []SignatoryAndAddress{}
_, _, err := surge.Unmarshal(&slice, data, surge.MaxBytes)
```

You can also adjust the `surge.MaxBytes` to ensure that the unmarshaller will not allocate too much memory.

File: peer/peerdiscovery.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+func (dc *DiscoveryClient) didReceivePing(from id.Signatory, msg wire.Msg) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var addressBuf [1024]byte
+	addressByteSlice := addressBuf[:0]
+	expNumPeer := int(binary.LittleEndian.Uint64(msg.Data[:8]))
+	port := uint16(binary.LittleEndian.Uint64(msg.Data[8:]))
+	ipAddr, _ := dc.transport.Table().IP(from)
```
Comment (loongy) [inline peer/peerdiscovery.go:108]: Make sure to capture the `ipAddrOk` as the second return value and handle the erroneous (and unexpected) case were it is false. Do not assume others won't modify the code in such a way that your assumption here (that the second return value is always true) will hold.

**Review by loongy — CHANGES_REQUESTED** [review]

File: peer/peerdiscovery.go (hunk 4 above)
Comment (rahulghangas) [reply peer/peerdiscovery.go]: This doesn't seem to work

File: peer/peer.go (hunk 1 above)
Comment (rahulghangas) [reply peer/peer.go]: This will actually cause problems. Peer discovery can only be run if `p.transport.Run` has started accepting messages. Starting them at the same time will cause peer discovery to throw errors. Secondly, a user might be bootstrapping and might not want to run peer discovery anyway. I think we should leave the responsibility on the user

File: dht/table.go [hunk 6]
```diff
@@ -26,6 +26,9 @@ type Table interface {
 	// PeerAddress returns the network address associated with the given peer.
 	PeerAddress(id.Signatory) (wire.Address, bool)
 
+	RegisterIP(id.Signatory, string)
+	IP(id.Signatory) (string, bool)
```
Comment (loongy) [inline dht/table.go:30]: We also need a method for deleting IP addresses, otherwise it will grow unbounded.

Also, for consistency, we should probably call this `AddIP` as oppose to `RegisterIP`. The functionality is no different to that of `AddPeer` (but for IPs), so it is odd that it has a different naming convention.

File: peer/peer.go [hunk 7]
```diff
@@ -28,6 +28,7 @@ type Peer struct {
 	transport *transport.Transport
 	syncer    *Syncer
 	gossiper  *Gossiper
+	discoveryClient *DiscoveryClient
```
Comment (loongy) [inline peer/peer.go:31]: Please run `gofmt` to make sure all spacing/formatting is standard.

File: peer/peer.go [hunk 8]
```diff
@@ -80,6 +82,10 @@ func (p *Peer) Gossip(ctx context.Context, contentID []byte, subnet *id.Hash) {
 	p.gossiper.Gossip(ctx, contentID, subnet)
 }
 
+func (p *Peer) PeerDiscovery(ctx context.Context) {
```
Comment (loongy) [inline peer/peer.go:85]: Methods are conventionally named as verbs. This should be `DiscoverPeers` (likewise for the method on the actual client object too).

File: peer/peerdiscovery.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	"time"
+
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	"go.uber.org/zap"
+)
+
+type SignatoryAndAddress struct {
```
Comment (loongy) [inline peer/peerdiscovery.go:16]: This type definition belongs in the `wire` package.

File: peer/peerdiscovery.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+	var pingData [2]byte
+	binary.LittleEndian.PutUint16(pingData[:], dc.transport.Port())
+
+	msg := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
```
Comment (loongy) [inline peer/peerdiscovery.go:68]: Tickers must be stopped using `defer ticker.Stop()` otherwise resources are leaked.

File: peer/peerdiscovery.go [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+	return nil
+}
+
+func (dc *DiscoveryClient) didReceivePing(from id.Signatory, msg wire.Msg) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var addressBuf [1024]byte
+	addressByteSlice := addressBuf[:0]
+	port := binary.LittleEndian.Uint16(msg.Data)
```
Comment (loongy) [inline peer/peerdiscovery.go:105]: Before doing this, you must check that `msg.Data` has a length of exactly 2 bytes (otherwise the ping data should be considered malformed).

File: peer/peerdiscovery.go [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+	ipAddr, ipAddrOk := dc.transport.Table().IP(from)
+	if !ipAddrOk {
+		return fmt.Errorf("ip address for remote peer not found")
```
Comment (loongy) [inline peer/peerdiscovery.go:108]: Instead of returning an error immediately, you should check if the peer already has a known `wire.Address` in the table. If so, then return nil, instead of returning an error.

File: peer/peerdiscovery.go [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	dc.transport.Table().AddPeer(
+		from,
+		wire.NewUnsignedAddress(wire.TCP, fmt.Sprintf("%v:%v", ipAddr, port), uint64(time.Now().UnixNano())),
+		)
```
Comment (loongy) [inline peer/peerdiscovery.go:113]: Again, run `gofmt`.

Also, this is probably where you would delete the IP from the table.

File: peer/peerdiscovery.go [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+	peers := dc.transport.Table().Peers(dc.opts.MaxExpectedPeers)
+	for _, sig := range peers {
+		var addrAndSig [128]byte
```
Comment (loongy) [inline peer/peerdiscovery.go:117]: This only leaves 64 bytes for the network address. It is completely feasible that more than 64 bytes would be needed. Increasing this to 256 bytes gives us a more reasonable amount to work with.

File: peer/peerdiscovery.go [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+		if err != nil {
+			dc.opts.Logger.Debug("sending pingAck", zap.Error(err))
+		}
+		addressByteSlice = append(addressByteSlice, addrAndSigSlice[:len(addrAndSigSlice)-len(tail)]...)
+	}
+	response := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePingAck,
+		To:      id.Hash(from),
+		Data:    addressByteSlice,
```
Comment (loongy) [inline peer/peerdiscovery.go:134]: Instead of doing this (and the complex marshalling in the above for loop) you should instead define `[]SignatoryAndAddress`, append to that, and then marshal it once at the end. Surge can handle slices very well.

File: peer/peerdiscovery.go [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+		addr, ok := dc.transport.Table().PeerAddress(sig)
+		if !ok {
+			dc.opts.Logger.Debug("sending pingAck", zap.String("peer", "does not exist in table"))
+		}
+		sigAndAddr := SignatoryAndAddress{Signatory: sig, Address: addr}
+		tail, _, err := sigAndAddr.Marshal(addrAndSigSlice, len(addrAndSigSlice))
+		if err != nil {
+			dc.opts.Logger.Debug("sending pingAck", zap.Error(err))
+		}
+		addressByteSlice = append(addressByteSlice, addrAndSigSlice[:len(addrAndSigSlice)-len(tail)]...)
```
Comment (loongy) [inline peer/peerdiscovery.go:128]: This is convoluted and requires a lot of marshalling. Consider using a `[]SignatoryAndAddress` instead of `[]byte` and just marshal it once after the loop.

File: peer/peerdiscovery.go [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	return nil
+}
+
+func (dc *DiscoveryClient) didReceivePingAck(from id.Signatory, msg wire.Msg) error {
+	var sigAndAddr SignatoryAndAddress
+	var sigAndAddrArray [20]SignatoryAndAddress
+	sigAndAddrSlice := sigAndAddrArray[:0]
+
+	if cap(sigAndAddrArray) > dc.opts.MaxExpectedPeers {
```
Comment (loongy) [inline peer/peerdiscovery.go:148]: You create the cap yourself on the line above, so this check makes no sense.

File: peer/peerdiscovery.go [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+	err := dc.transport.Send(ctx, from, response)
+	if err != nil {
+		dc.opts.Logger.Debug("sending pingAck", zap.Error(err))
+	}
+	return nil
+}
+
+func (dc *DiscoveryClient) didReceivePingAck(from id.Signatory, msg wire.Msg) error {
+	var sigAndAddr SignatoryAndAddress
+	var sigAndAddrArray [20]SignatoryAndAddress
```
Comment (loongy) [inline peer/peerdiscovery.go:145]: No need to use a static array. Just leave it as an empty slice.

File: peer/peerdiscovery.go [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+	sigAndAddrSlice := sigAndAddrArray[:0]
+
+	if cap(sigAndAddrArray) > dc.opts.MaxExpectedPeers {
+		sigAndAddrSlice = make([]SignatoryAndAddress, 0, dc.opts.MaxExpectedPeers)
+	}
+
+	dataLeft := msg.Data
+	count := 0
+	for len(dataLeft) > 0 {
```
Comment (loongy) [inline peer/peerdiscovery.go:154]: Again, if you initially marshal `[]SignatoryAndAddress` (as suggested above) then you can just unmarshal into `sigAndAddrSlice` in 3 LoC (include error checking) instead of using this complicated loop.

File: transport/transport.go [hunk 20]
```diff
@@ -337,14 +356,10 @@ func (t *Transport) dial(retryCtx context.Context, remote id.Signatory, remoteAd
 			case <-retryCtx.Done():
 			case <-dialCtx.Done():
 				if !t.IsConnected(remote) {
-					// Cancel current dial context if restarting loop
-					cancel()
```
Comment (loongy) [inline transport/transport.go:341]: Why was this removed?

File: transport/transport.go [hunk 21]
```diff
@@ -337,14 +356,10 @@ func (t *Transport) dial(retryCtx context.Context, remote id.Signatory, remoteAd
...
 					continue
 				}
 			}
 		}
-		// Cancel last dial context before exiting
-		cancel()
```
Comment (loongy) [inline transport/transport.go:347]: Why was this removed?

**Review by loongy — CHANGES_REQUESTED** [review]

File: peer/peerdiscovery.go (hunk 15 above)
Comment (rahulghangas) [reply peer/peerdiscovery.go:135]: I'm a bit confused about marshalling and unmarshalling. While the sizehint for a `SignatoryAndAddress` struct return 110 bytes, the actual size in bytes of the unmarshalled slice is as follows
```
for x elements in slice
size -> 4 + 14 * x + 110 * x
```

I understand the initial 4 bytes, but where did the 14 extra bytes come from?

File: transport/transport.go (hunk 20 above)
Comment (rahulghangas) [reply transport/transport.go:341]: I was removing the `defer cancel()` in the loop and replacing it with specific cancel calls

File: transport/transport.go (hunk 21 above)
Comment (rahulghangas) [reply transport/transport.go:347]: I was removing the `defer cancel()` in the loop and replacing it with specific cancel calls

File: dht/table.go [hunk 22]
```diff
@@ -93,12 +107,14 @@ func (table *InMemTable) AddPeer(peerID id.Signatory, peerAddr wire.Address) {
 
 	// Insert into the sorted address list based on its XOR distance from our
 	// own address.
-	i := sort.Search(len(table.sorted), func(i int) bool {
-		return table.isCloser(peerID, table.sorted[i])
-	})
-	table.sorted = append(table.sorted, id.Signatory{})
-	copy(table.sorted[i+1:], table.sorted[i:])
-	table.sorted[i] = peerID
+	if !ok {
```
Comment (loongy) [inline dht/table.go:110]: Why is this condition here? It prevents the address of a peer from ever being updated. I'm not sure that's the behaviour we want.

File: peer/peerdiscovery.go [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
+	defer ticker.Stop()
+	for {
+		for _, sig := range dc.transport.Table().Peers(dc.opts.Alpha) {
+			msg.To = id.Hash(sig)
+			err := dc.transport.Send(ctx, sig, msg)
```
Comment (loongy) [inline peer/peerdiscovery.go:46]: This `ctx` should be scoped. That is, we should create a wrapping context that has a smaller timeout, because `ctx` as defined at this point is the _entire lifetime_ of the peer discovery loop. This is typically going to be the entire lifetime of the application. Obviously way too high for a single request.

**Review by loongy — CHANGES_REQUESTED** [review]

File: peer/peerdiscovery.go (hunk 15 above)
Comment (loongy) [reply peer/peerdiscovery.go]: I believe it comes from the recursive nature of how marshalling in slices works. But the couple of extra bytes is definitely worth the safety/simplicity of implementation.

File: dht/table.go (hunk 22 above)
Comment (rahulghangas) [reply dht/table.go:110]: This condition only stops re-adding a signatory to the sorted array of signatories, maintaining uniqueness of elements. The address can be replaced regardless and is done a couple lines before this conditional statement.

File: dht/table.go [hunk 24]
```diff
@@ -26,6 +26,13 @@ type Table interface {
 	// PeerAddress returns the network address associated with the given peer.
 	PeerAddress(id.Signatory) (wire.Address, bool)
 
+	// AddPeer to the table with an associate network address.
+	AddIP(id.Signatory, string)
+	// DeleteIP from the table.
+	DeleteIP(id.Signatory)
+	// IP returns the network ip address associated with the given peer.
+	IP(id.Signatory) (string, bool)
```
Comment (loongy) [inline dht/table.go:34]: You need to add tests for these methods.

File: dht/table.go [hunk 25]
```diff
@@ -93,12 +107,14 @@ func (table *InMemTable) AddPeer(peerID id.Signatory, peerAddr wire.Address) {
...
-	table.sorted = append(table.sorted, id.Signatory{})
-	copy(table.sorted[i+1:], table.sorted[i:])
-	table.sorted[i] = peerID
+	if !ok {
+		i := sort.Search(len(table.sorted), func(i int) bool {
+			return table.isCloser(peerID, table.sorted[i])
+		})
+		table.sorted = append(table.sorted, id.Signatory{})
+		copy(table.sorted[i+1:], table.sorted[i:])
+		table.sorted[i] = peerID
```
Comment (loongy) [inline dht/table.go:116]: You need to add a test for `AddPeer` that verifies this behaviour works as intended (e.g. that the peer can be updated, but its position in the sorted list does not change).

File: peer/sync.go [hunk 26]
```diff
@@ -82,6 +82,13 @@ func (syncer *Syncer) Sync(ctx context.Context, contentID []byte, hint *id.Signa
 	}
 	syncer.pendingMu.Unlock()
 
+	// Ensure that pending content is removed.
+	defer func() {
+		syncer.pendingMu.Lock()
+		delete(syncer.pending, string(contentID))
+		syncer.pendingMu.Unlock()
+	}()
```
Comment (loongy) [inline peer/sync.go:90]: You need to add a test that fails if this code is removed, but passes if this code is present. Otherwise, we cannot be sure that this has actually fixed the problem it is supposed to fix.

File: transport/transport.go [hunk 27]
```diff
@@ -209,6 +223,8 @@ func (t *Transport) run(ctx context.Context) {
 		}
 	}()
 
+	t.table.AddPeer(t.self, wire.NewUnsignedAddress(wire.TCP, fmt.Sprintf("%v:%v", t.opts.Host, t.opts.Port), uint64(time.Now().UnixNano())))
```
Comment (loongy) [inline transport/transport.go:226]: You need to add a test (or check to an existing test) that demonstrates that our own unsigned address is added/updated when the transport is run. This ensures future modifications do not break the assumption that this is being done.

File: transport/transport.go [hunk 28]
```diff
@@ -225,6 +241,8 @@ func (t *Transport) run(ctx context.Context) {
 			enc = codec.LengthPrefixEncoder(codec.PlainEncoder, enc)
 			dec = codec.LengthPrefixDecoder(codec.PlainDecoder, dec)
 
+			t.table.AddIP(remote, addr[:strings.IndexByte(addr, ':')])
```
Comment (loongy) [inline transport/transport.go:244]: You need to add a test (or check to an existing test) that demonstrates the transport table has the IP of the client after the client has connected / sent a message. This ensures future modifications do not break the assumption that this is being done.

File: peer/peerdiscovery.go [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+
+	alpha := dc.opts.Alpha
+	sendDuration := dc.opts.PingTimePeriod/time.Duration(alpha)
+	outer:
+	for {
+		for _, sig := range dc.transport.Table().Peers(alpha) {
+			innerCtx, cancel := context.WithTimeout(ctx, sendDuration)
+			msg.To = id.Hash(sig)
+			err := dc.transport.Send(innerCtx, sig, msg)
+			cancel()
```
Comment (loongy) [inline peer/peerdiscovery.go:51]:
```suggestion
			err := func() error {
				innerCtx, innerCancel := context.WithTimeout(ctx, sendDuration)
				defer innerCancel()
				msg.To = id.Hash(sig)
				return dc.transport.Send(innerCtx, sig, msg)
			}()
```
You should always try and use the `defer` statement, especially when making calls to functions, because it will ensure that the deferred statement is _always_ run (even if one of the function calls panics).

File: peer/peerdiscovery.go [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
+	defer ticker.Stop()
+
+	alpha := dc.opts.Alpha
+	sendDuration := dc.opts.PingTimePeriod/time.Duration(alpha)
+	outer:
```
Comment (loongy) [inline peer/peerdiscovery.go:45]: Nice use of outer loop labelling. This is an interesting way to restart the ping process if it exceeds the tick interval. Idiomatically, this label should be capitalised.

File: peer/peerdiscovery.go [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+
+	peers := dc.transport.Table().Peers(dc.opts.MaxExpectedPeers)
+	addrAndSig := make([]wire.SignatoryAndAddress, 0, len(peers))
+	for _, sig := range peers {
+		addr, ok := dc.transport.Table().PeerAddress(sig)
+		if !ok {
+			dc.opts.Logger.Debug("sending pingAck", zap.String("peer", "does not exist in table"))
+		}
+		sigAndAddr := wire.SignatoryAndAddress{Signatory: sig, Address: addr}
+		addrAndSig = append(addrAndSig, sigAndAddr)
```
Comment (loongy) [inline peer/peerdiscovery.go:115]: This is probably not what you intended. This `Debug` should be a `DPanic` and then a `continue`. Either way, I doubt you want to fall-through to the next statements. Consider adding a test that catches this mistake.

File: peer/peerdiscovery_test.go [hunk 32]
```diff
@@ -0,0 +1,66 @@
...
+	"time"
+
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Peer", func() {
```
Comment (loongy) [inline peer/peerdiscovery_test.go:15]:
```suggestion
var _ = Describe("Peer Discovery", func() {
```

File: peer/peerdiscovery_test.go [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+			for i := range peers {
+				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+				defer cancel()
+				go peers[i].Run(ctx)
+				tables[i].AddPeer(opts[(i+1)%n].PrivKey.Signatory(),
+					wire.NewUnsignedAddress(wire.TCP,
+						fmt.Sprintf("%v:%v", "localhost", uint16(3333+((i+1)%n))), uint64(time.Now().UnixNano())))
+			}
+
+			<-time.After(1 * time.Second)
```
Comment (loongy) [inline peer/peerdiscovery_test.go:33]:
```suggestion
			time.Sleep(time.Second)
```

File: peer/peerdiscovery_test.go [hunk 34]
```diff
@@ -0,0 +1,66 @@
...
+			for i := range peers {
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				go peers[i].DiscoverPeers(ctx)
+			}
+			<-time.After(4 * time.Second)
```
Comment (loongy) [inline peer/peerdiscovery_test.go:39]:
```suggestion
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			for i := range peers {
				go peers[i].DiscoverPeers(ctx)
			}
			<-ctx.Done()
```
It is better to rely on concrete moments in the test (e.g. "by the time the context has expired") than to rely on ad-hoc timeouts. This way, if someone extends the context duration, this test will still work as expected.

File: peer/peerdiscovery_test.go [hunk 35]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Peer", func() {
+	Context("when trying to discover other peers using the peer discovery client", func() {
+		It("should successfully find all peers", func() {
```
Comment (loongy) [inline peer/peerdiscovery_test.go:17]: This is a great end-to-end test. However, it only checks that discovery works as expected when the nodes start in a ring topology. Consider adding a wrapping `Context("when peers begin in a ring topology")` and then add a few more topologies to this test. What about a star topology? What about a linear topology (where the last peer is _not_ connected to the first one)? What about a fully connected by random topology? All of these are very important to test.

File: peer/peerdiscovery_test.go [hunk 36]
```diff
@@ -0,0 +1,66 @@
...
+						Expect(addr.Value).To(Or(
+							Equal(fmt.Sprintf("127.0.0.1:%v", uint16(3333+j))),
+							Equal(fmt.Sprintf("localhost:%v", uint16(3333+j))),
+							Equal(fmt.Sprintf(":%v", uint16(3333+j)))))
+					}
+				}
+			}
+		})
+	})
+})
```
Comment (loongy) [inline peer/peerdiscovery_test.go:66]: Consider adding some negative tests to check that misbehaving remote peers cannot cause crashes/panic or unexpected behaviour for our own peer. For example, what if they send malformed pings or ping acknowledgements? Really long ones? Empty ones? Random ones?

**Review by loongy — CHANGES_REQUESTED** [review]

File: peer/sync.go (hunk 26 above)
Comment (rahulghangas) [reply peer/sync.go:90]: Not sure how I would test this since the map itself is a private field

Comment (jazg) [reply peer/sync.go:90]: @rahulghangas You don't need to interact with the field. Just need a general syncing test (like the example I posted in Slack) that fails if that code is commented out, and passes when the code is added.

Comment (rahulghangas) [reply peer/sync.go:90]: The purpose of this block is to make sure that a pending struct is deleted once a sync message is successfully received or a sync request fails. To make sure that it is indeed deleted, I need to check the map for the corresponding `contentID` once the function is exited

Comment (rahulghangas) [reply peer/sync.go:90]: @loongy not sure how to test this

**Comment by rahulghangas** [comment]
@loongy needs another review

File: peer/sync.go (hunk 26 above)
Comment (jazg) [reply peer/sync.go:90]: @rahulghangas You could have a general test that fails without it and succeeds once you have it. For example:

1. N nodes in the network
2. Node A has some data
3. Node B attempts to sync this data
4. Node B’s context expires so it is unable to sync the data
5. Node B attempts to sync again, but it realises there is already a pending channel it can wait on
6. Node B decides to wait on this channel, and never receives the content because the original sync attempt has already expired

This test would fail without the pending map clean-up, but should pass once you add it.
